crates/turborepo-lib/src/run/watch.rs

```diff
@@ -1,37 +1,231 @@
+        }
+
+        let selectors = filters
+            .iter()
+            .map(|f| TargetSelector::from_str(f))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        for selector in selectors {
+            if !selector.from_ref.is_empty() || !selector.to_ref_override.is_empty() {
+                let filter = format!("--filter={}", selector.raw);
+                let span: SourceSpan = (0, filter.len()).into();
+
+                return Err(Error::GitRangeInFilter { filter, span });
+            }
+        }
+
+        Ok(())
+    }
+    pub async fn start(base: CommandBase, telemetry: CommandEventBuilder) -> Result<(), Error> {
+        let signal = commands::run::get_signal()?;
+        let handler = SignalHandler::new(signal);
+        let Some(subscriber) = handler.subscribe() else {
+            tracing::warn!("failed to subscribe to signal handler, shutting down");
+            return Ok(());
+        };
+
+        let Some(Command::Watch(args)) = &base.args().command else {
+            unreachable!();
+        };
+
+        let opts = Opts::try_from(base.args())?;
+
+        Self::validate_filter(&opts, &args.filter)?;
+        // We currently don't actually need the whole Run struct, just the filtered
+        // packages. But in the future we'll likely need it to more efficiently
+        // spawn tasks.
+        let mut run = RunBuilder::new(base.clone())?
+            .build(&handler, telemetry.clone())
+            .await?;
+
         let connector = DaemonConnector {
             can_start_server: true,
             can_kill_server: true,
-            paths: DaemonPaths::from_repo_root(repo_root),
+            paths: DaemonPaths::from_repo_root(&base.repo_root),
         };
 
         let mut client = connector.connect().await?;
 
-        let mut hashes = client.package_changes().await?;
-        while let Some(hash) = hashes.next().await {
-            // Should we recover here?
-            let hash = hash.unwrap();
-            let event = hash.event.expect("event is missing");
-            match event {
-                proto::package_change_event::Event::PackageChanged(proto::PackageChanged {
-                    package_name,
-                }) => {
-                    println!("{} changed", package_name);
+        let mut events = client.package_changes().await?;
+        let mut current_runs: HashMap<PackageName, JoinHandle<Result<i32, run::Error>>> =
+            HashMap::new();
+        let event_fut = async {
+            while let Some(event) = events.next().await {
+                let event = event.unwrap();
+                Self::handle_change_event(
+                    &mut run,
+                    event.event.unwrap(),
+                    &mut current_runs,
+                    &base,
+                    &telemetry,
+                    &handler,
+                )
+                .await?;
+            }
+
+            Ok::<(), Error>(())
+        };
+
+        select! {
```

Comment:
I think this should be `biased;`, preferring the signal handler. Don't want shutdown blocked on a flood of changes.